### DISCUSSION TITLE
Make it easier to perform bitwise operations between integer values and BIT literals

### DISCUSSION THREAD

**Discussion by rpbouman** [duckdb/duckdb#14999] [Ideas]
Request: allow operations like:

intvalue & bitstring-literal

Usage
When you have integer data with bitfield semantics, and need to extract bits from it, then it would be really nice if we could write that explicitly using `BIT` literals. 

(Concrete use case for me is controlling the ordering of super-aggregate rows in queries that use `GROUP BY GROUPING SETS`)

I would like to write something like:

`intvalue & '010'::BIT`

But this yields:

> Binder Error: No function matches the given name and argument types '&(INTEGER_LITERAL, BIT)'. You might need to add explicit type casts.

Unfortunately, an explicit cast of the intvalue won't do:

`intvalue::BIT & '010'::BIT`

> Invalid Input Error: Cannot AND bit strings of different sizes

This would be fine if we could write something like:

`intvalue::BIT(3) & '010'::BIT`

but then we get: 

> Parser Error: Type `BIT `does not support any modifiers!

Casting the BIT to an int does work:

`intvalue::BIT & '010'::BIT::INT`

but this feels a bit strange, as if we're taking an extra step.

Possible features that would contribute to helping with this issue:
- Let the `BIT `datatype have a length modifier so that we could write `intvalue::BIT(3)`. In addition, the corresponding cast operation should not consider it to be an error if the number of bits of the integer type exceeds the length of the bitstring. Regardless of whether the aim is to make int/bit operations easier, I think this modifier would make sense considering that now already the length of the `BIT` values is checked in operations. It already feels like `BIT` values of different length are really distinct types, it's just that we don't have a direct way of expressing and manipulating that.
- Implement implicit casting of integer-type values to `BIT` values. This should be applied when applying bitwise operators to an operand of the integer-family and a `BIT`-operand. The `BIT` value resulting from the cast of the integer value should have the same length as the `BIT`-operand. So if the number of bits in the integer value exceeds the length of the BIT-value, the integer bits would be truncated; If the length of the BIT value exceeds the number of bits in the integer value, then BIT resulting from the cast of the integer value should be zero-padded to match the length of the BIT value

**Comment by soerenwolfers** [comment]
FYI There is also a `bitstring(b, length)` function that pads bitstrings on the left. Might not satisfy your hope for more succinct operations though. 

What type do you think `int & bitstring` should have?

**Reply by rpbouman** [reply]
Hi @soerenwolfers, thanks for chiming in! 

>FYI There is also a bitstring(b, length) function that pads bitstrings on the left. Might not satisfy your hope for more succinct operations though.

Yes, I know about that function. The problem I see with that one is that it requires you to know the exact type (TINY, SMALL, BIG, HUGE) of the integer value, and then to figure out what length the bitstring should be. Something that can certainly be solved but it won't be very pretty or concise.

For the example I mentioned, it's the other way around: I know what groupings appear in the GROUPING SETS clause, and I know how many distinct columns are used for grouping. So I know exactly how many useful bits the GROUPING_ID() function will return, and I think it's totally reasonable to use that information directly in the comparison. 

WIth `bitstring(x, length)`, we're turning things upside down: I have to "upgrade" the number of columns to have just as many bits as whatever particular flavor of integer the GROUPING_ID function returns.

> What type do you think int & bitstring should have?

In my mind, the type should be BITSTRING, and it probably shouldn't matter whether the BITSTRING is left or right of the operator. If you're doing this kind of operations, then it's because the integer is really a bitfield. If it were the other way around, and your bitfield value was really a number, then you'd be casting the bitfield to integer and then do your operation on those two integer operands.
